File: #153_Find_Minimum_in_Rotated_Sorted_Array/Solution.java
public class Solution {
    public int findMin(int[] nums) {
        if (nums.length == 0) return 0;
        int lo = 0, hi = nums.length - 1;
        while (true) {
        	int mid = (lo + hi) / 2;
	        if (nums[lo] <= nums[mid] && nums[mid] <= nums[hi]) return nums[lo];
        	if (nums[lo] <= nums[mid]) lo = mid + 1;
        	else if (nums[mid] < nums[hi]) hi = mid;
        }
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {4, 5, 6, 7, 0, 1, 2};
    	System.out.println(sol.findMin(t));
    }
}

File: #153_Find_Minimum_in_Rotated_Sorted_Array/Solution.py
'''
time: O(log(n))
'''

class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def findMin(self, nums):
        if len(nums) == 1:
        	return nums[0]
        if len(nums) == 2:
        	return min(nums)
        if nums[0] < nums[len(nums) - 1]:
        	return nums[0]
        mid = (len(nums) - 1) / 2
        if nums[mid] > nums[0]:
        	return self.findMin(nums[mid + 1:])
        if nums[mid] < nums[0]:
        	return self.findMin(nums[:mid + 1])

sol = Solution()
print sol.findMin([2, 3, 4, 5, 0, 1])